File: src/main/java/net/sf/rails/game/PhaseManager.java
package net.sf.rails.game;

import java.util.List;
import java.util.Map;

import net.sf.rails.common.parser.Configurable;
import net.sf.rails.common.parser.ConfigurationException;
import net.sf.rails.common.parser.Tag;
import net.sf.rails.game.state.GenericState;
import net.sf.rails.game.state.Owner;
import net.sf.rails.game.state.State;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;


public class PhaseManager extends RailsManager implements Configurable {
    // static data
    private final List<Phase> phaseList = Lists.newArrayList();
    private final Map<String, Phase> phaseMap = Maps.newHashMap();
    
    // dynamic data
    private final GenericState<Phase> currentPhase = GenericState.create(this, "currentPhase");

    /**
     * Used by Configure (via reflection) only
     */
    public PhaseManager(RailsRoot parent, String id) {
        super(parent, id);
    }
    
    public void configureFromXML(Tag tag) throws ConfigurationException {
        /*
         * Phase class name is now fixed but can be made configurable, if
         * needed.
         */
        List<Tag> phaseTags = tag.getChildren("Phase");
        Phase phase;
        Phase previousPhase = null;
        String name;

        int n = 0;
        for (Tag phaseTag : phaseTags) {
            name = phaseTag.getAttributeAsString("name", String.valueOf(n + 1));
            phase = new Phase(this, name, n++, previousPhase);
            phaseList.add(phase);
            phaseMap.put(name, phase);
            phase.configureFromXML(phaseTag);
            previousPhase = phase;
        }
    }
    
    public void finishConfiguration (RailsRoot root) 
    throws ConfigurationException {
        
        for (Phase phase : phaseList) {
            phase.finishConfiguration(root);
        }
        
        Phase initialPhase = phaseList.get(0);
        setPhase(initialPhase, null);
    }

    public Phase getCurrentPhase() {
        return currentPhase.value();
    }
    
    public State getCurrentPhaseModel() {
        return currentPhase;
    }

    public int getCurrentPhasendex() {
        return getCurrentPhase().getIndex();
    }

    public void setPhase(String name, Owner lastTrainBuyer) {
        setPhase(phaseMap.get(name), lastTrainBuyer);
    }

    public void setPhase(Phase phase, Owner lastTrainBuyer) {
        if (phase != null) {
            phase.setLastTrainBuyer (lastTrainBuyer);
            currentPhase.set(phase);
            phase.activate();
        }
    }

    public Phase getPhaseByName(String name) {
        return phaseMap.get(name);
    }

    public boolean hasReachedPhase(String phaseName) {
        return getCurrentPhase().getIndex() >= getPhaseByName(phaseName).getIndex();

    }

    public List<Phase> getPhases() {
        return phaseList;
    }
    
}

File: dotnet/RailsLib.Net/Net/Game/PhaseManager.cs
using GameLib.Net.Common.Parser;
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Text;

namespace GameLib.Net.Game
{
    public class PhaseManager : RailsManager, IConfigurable
    {
        // static data
        private List<Phase> phaseList = new List<Phase>();
        private Dictionary<string, Phase> phaseMap = new Dictionary<string, Phase>();

        // dynamic data
        private GenericState<Phase> currentPhase;

        /**
         * Used by Configure (via reflection) only
         */
        public PhaseManager(RailsRoot parent, string id) : base(parent, id)
        {
            currentPhase = GenericState<Phase>.Create(this, "currentPhase");
        }

        public void ConfigureFromXML(Tag tag)
        {
            /*
             * Phase class name is now fixed but can be made configurable, if
             * needed.
             */
            List<Tag> phaseTags = tag.GetChildren("Phase");
            Phase phase;
            Phase previousPhase = null;
            string name;

            int n = 0;
            foreach (Tag phaseTag in phaseTags)
            {
                name = phaseTag.GetAttributeAsString("name", $"{n + 1}");
                phase = new Phase(this, name, n++, previousPhase);
                phaseList.Add(phase);
                phaseMap[name] = phase;
                phase.ConfigureFromXML(phaseTag);
                previousPhase = phase;
            }
        }

        public void FinishConfiguration(RailsRoot root)
        {
            foreach (Phase phase in phaseList)
            {
                phase.FinishConfiguration(root);
            }

            Phase initialPhase = phaseList[0];
            SetPhase(initialPhase, null);
        }

        public Phase GetCurrentPhase()
        {
            return currentPhase.Value;
        }

        public GameState CurrentPhaseModel
        {
            get
            {
                return currentPhase;
            }
        }

        public int CurrentPhaseIndex
        {
            get
            {
                return GetCurrentPhase().Index;
            }
        }

        public void SetPhase(string name, IOwner lastTrainBuyer)
        {
            SetPhase(phaseMap[name], lastTrainBuyer);
        }

        public void SetPhase(Phase phase, IOwner lastTrainBuyer)
        {
            if (phase != null)
            {
                phase.SetLastTrainBuyer(lastTrainBuyer);
                currentPhase.Set(phase);
                phase.Activate();
            }
        }

        public Phase GetPhaseByName(string name)
        {
            return phaseMap[name];
        }

        public bool HasReachedPhase(string phaseName)
        {
            return GetCurrentPhase().Index >= GetPhaseByName(phaseName).Index;

        }

        public List<Phase> Phases
        {
            get
            {
                return phaseList;
            }
        }
    }
}
